File: packages/ts/server/src/storage/s3.ts
import {
  S3Client,
  GetObjectCommand,
  PutObjectCommand,
  ListObjectsV2Command,
  DeleteObjectCommand,
  DeleteObjectsCommand,
} from "@aws-sdk/client-s3"
import type { ObjectStore, StoreContext } from "./base.js"

export interface S3StorageOptions {
  accessKeyId: string
  secretAccessKey: string
  /** S3-compatible endpoint URL (e.g. "https://s3.amazonaws.com" or a MinIO URL). */
  endpoint: string
  bucket: string
  region?: string
  /**
   * Use path-style addressing (required for most S3-compatible services such as MinIO).
   * Defaults to `true`.
   */
  forcePathStyle?: boolean
}

export class S3ObjectStore implements ObjectStore {
  private _client: S3Client
  private _bucket: string

  constructor(opts: S3StorageOptions) {
    this._bucket = opts.bucket
    this._client = new S3Client({
      endpoint: opts.endpoint,
      region: opts.region ?? "us-east-1",
      credentials: {
        accessKeyId: opts.accessKeyId,
        secretAccessKey: opts.secretAccessKey,
      },
      forcePathStyle: opts.forcePathStyle ?? true,
    })
  }

  async getString(key: string, _context?: StoreContext): Promise<string | null> {
    try {
      const resp = await this._client.send(new GetObjectCommand({ Bucket: this._bucket, Key: key }))
      return await resp.Body!.transformToString("utf-8")
    } catch (e: any) {
      if (e.name === "NoSuchKey") return null
      throw e
    }
  }

  async put(
    key: string,
    body: string,
    opts?: { contentType?: string; cacheControl?: string },
    _context?: StoreContext,
  ): Promise<void> {
    await this._client.send(
      new PutObjectCommand({
        Bucket: this._bucket,
        Key: key,
        Body: body,
        ...(opts?.contentType != null && { ContentType: opts.contentType }),
        ...(opts?.cacheControl != null && { CacheControl: opts.cacheControl }),
      }),
    )
  }

  async getBytes(key: string, _context?: StoreContext): Promise<{ body: Uint8Array; contentType: string } | null> {
    try {
      const resp = await this._client.send(new GetObjectCommand({ Bucket: this._bucket, Key: key }))
      const contentType = resp.ContentType ?? "application/octet-stream"
      const body = await resp.Body!.transformToByteArray()
      return { body, contentType }
    } catch (e: any) {
      if (e.name === "NoSuchKey") return null
      throw e
    }
  }

  async putBytes(
    key: string,
    body: Uint8Array,
    opts: { contentType: string; cacheControl?: string },
    _context?: StoreContext,
  ): Promise<void> {
    await this._client.send(
      new PutObjectCommand({
        Bucket: this._bucket,
        Key: key,
        Body: body,
        ContentType: opts.contentType,
        ...(opts.cacheControl != null && { CacheControl: opts.cacheControl }),
      }),
    )
  }

  async listKeys(
    prefix: string,
    opts?: { startAfter?: string; limit?: number },
    _context?: StoreContext,
  ): Promise<string[]> {
    // S3 returns at most 1000 keys per page. Follow the continuation token so
    // the full key set is returned — the segmented append-only log keys ALL
    // chunks of a single document via `listKeys` (no `limit`), so a truncated
    // first page would silently drop every chunk past the 1000th and the
    // checkpoint bisect would read incomplete data. With a `limit` we stop as
    // soon as it is satisfied. (`StartAfter` is honored only on the first
    // request; the continuation token governs subsequent pages.)
    const keys: string[] = []
    let continuationToken: string | undefined
    do {
      const resp = await this._client.send(
        new ListObjectsV2Command({
          Bucket: this._bucket,
          Prefix: prefix,
          ...(continuationToken == null &&
            opts?.startAfter != null && { StartAfter: opts.startAfter }),
          ...(opts?.limit != null && { MaxKeys: opts.limit }),
          ...(continuationToken != null && { ContinuationToken: continuationToken }),
        }),
      )
      for (const obj of resp.Contents ?? []) keys.push(obj.Key!)
      if (opts?.limit != null && keys.length >= opts.limit) {
        return keys.slice(0, opts.limit)
      }
      continuationToken = resp.IsTruncated ? resp.NextContinuationToken : undefined
    } while (continuationToken != null)
    return keys
  }

  async delete(key: string, _context?: StoreContext): Promise<void> {
    await this._client.send(new DeleteObjectCommand({ Bucket: this._bucket, Key: key }))
  }

  async deleteMany(keys: string[], _context?: StoreContext): Promise<void> {
    if (keys.length === 0) return
    await this._client.send(
      new DeleteObjectsCommand({
        Bucket: this._bucket,
        Delete: { Objects: keys.map((k) => ({ Key: k })) },
      }),
    )
  }

  /** Destroy the underlying S3 client and release its resources. */
  destroy(): void {
    this._client.destroy()
  }
}

File: packages/python/server/starfish_server/storage/s3.py
"""S3-compatible object store implementation using aiobotocore."""


import asyncio
from dataclasses import dataclass
from typing import Any
from starfish_server.storage.base import AbstractObjectStore, StoreContext


@dataclass
class S3StorageOptions:
    """Configuration for the S3 object store."""

    access_key_id: str
    secret_access_key: str
    endpoint: str
    bucket: str
    region: str = "us-east-1"


class S3ObjectStore(AbstractObjectStore):
    """S3-compatible object store using aiobotocore."""

    def __init__(self, opts: S3StorageOptions) -> None:
        try:
            from aiobotocore.session import get_session
        except ImportError:
            raise ImportError(
                "aiobotocore is required for S3 storage. "
                "Install it with: pip install starfish-server[s3]"
            )
        self._session = get_session()
        self._opts = opts
        self._client_ctx: Any = None
        self._client: Any = None
        self._init_lock = asyncio.Lock()

    async def _get_client(self) -> Any:
        async with self._init_lock:
            if self._client is None:
                ctx = self._session.create_client(
                    "s3",
                    endpoint_url=self._opts.endpoint,
                    region_name=self._opts.region,
                    aws_access_key_id=self._opts.access_key_id,
                    aws_secret_access_key=self._opts.secret_access_key,
                )
                try:
                    self._client = await ctx.__aenter__()
                    self._client_ctx = ctx
                except BaseException:
                    self._client_ctx = None
                    raise
        return self._client

    async def close(self) -> None:
        """Close the underlying S3 client."""
        if self._client_ctx is not None:
            await self._client_ctx.__aexit__(None, None, None)
            self._client = None
            self._client_ctx = None

    async def get_string(self, key: str, *, context: StoreContext | None = None) -> str | None:  # noqa: ARG002
        client = await self._get_client()
        try:
            resp = await client.get_object(Bucket=self._opts.bucket, Key=key)
            body = await resp["Body"].read()
            return body.decode("utf-8")
        except client.exceptions.NoSuchKey:
            return None

    async def put(
        self,
        key: str,
        body: str,
        *,
        content_type: str | None = None,
        cache_control: str | None = None,
        context: StoreContext | None = None,  # noqa: ARG002
    ) -> None:
        client = await self._get_client()
        kwargs: dict[str, Any] = {
            "Bucket": self._opts.bucket,
            "Key": key,
            "Body": body.encode("utf-8"),
        }
        if content_type:
            kwargs["ContentType"] = content_type
        if cache_control:
            kwargs["CacheControl"] = cache_control
        await client.put_object(**kwargs)

    async def get_bytes(self, key: str, *, context: StoreContext | None = None) -> tuple[bytes, str] | None:  # noqa: ARG002
        client = await self._get_client()
        try:
            resp = await client.get_object(Bucket=self._opts.bucket, Key=key)
            body = await resp["Body"].read()
            content_type = resp.get("ContentType", "application/octet-stream")
            return body, content_type
        except client.exceptions.NoSuchKey:
            return None

    async def put_bytes(
        self,
        key: str,
        body: bytes,
        *,
        content_type: str,
        cache_control: str | None = None,
        context: StoreContext | None = None,  # noqa: ARG002
    ) -> None:
        client = await self._get_client()
        kwargs: dict[str, Any] = {
            "Bucket": self._opts.bucket,
            "Key": key,
            "Body": body,
            "ContentType": content_type,
        }
        if cache_control:
            kwargs["CacheControl"] = cache_control
        await client.put_object(**kwargs)

    async def list_keys(
        self,
        prefix: str,
        *,
        start_after: str | None = None,
        limit: int | None = None,
        context: StoreContext | None = None,  # noqa: ARG002
    ) -> list[str]:
        # S3 returns at most 1000 keys per page. Follow the continuation token
        # so the full key set is returned — the segmented append-only log keys
        # ALL chunks of a single document via ``list_keys`` (no ``limit``), so a
        # truncated first page would silently drop every chunk past the 1000th
        # and the checkpoint bisect would read incomplete data. With a ``limit``
        # we stop as soon as it is satisfied. (``StartAfter`` is honored only on
        # the first request; the continuation token governs subsequent pages.)
        client = await self._get_client()
        keys: list[str] = []
        continuation_token: str | None = None
        while True:
            kwargs: dict[str, Any] = {
                "Bucket": self._opts.bucket,
                "Prefix": prefix,
            }
            if continuation_token is None and start_after:
                kwargs["StartAfter"] = start_after
            if limit:
                kwargs["MaxKeys"] = limit
            if continuation_token is not None:
                kwargs["ContinuationToken"] = continuation_token

            resp = await client.list_objects_v2(**kwargs)
            for obj in resp.get("Contents", []):
                keys.append(obj["Key"])
            if limit and len(keys) >= limit:
                return keys[:limit]
            if not resp.get("IsTruncated"):
                return keys
            continuation_token = resp.get("NextContinuationToken")
            if continuation_token is None:
                return keys

    async def delete(self, key: str, *, context: StoreContext | None = None) -> None:  # noqa: ARG002
        client = await self._get_client()
        await client.delete_object(Bucket=self._opts.bucket, Key=key)

    async def delete_many(self, keys: list[str], *, context: StoreContext | None = None) -> None:  # noqa: ARG002
        if not keys:
            return
        client = await self._get_client()
        await client.delete_objects(
            Bucket=self._opts.bucket,
            Delete={"Objects": [{"Key": k} for k in keys]},
        )
